org.osgi.test.common/src/main/java/org/osgi/test/common/CloseableBundleContext.java

```diff
@@ -0,0 +1,239 @@
+			} else if (listener instanceof FrameworkListener) {
+				bundleContext.removeFrameworkListener((FrameworkListener) listener);
+			}
+		});
+	}
+
+	@Override
+	public String getProperty(String key) {
+		return bundleContext.getProperty(key);
+	}
+
+	@Override
+	public Bundle getBundle() {
+		return bundleContext.getBundle();
+	}
+
+	@Override
+	public Bundle installBundle(String location, InputStream input) throws BundleException {
+		Bundle bundle = bundleContext.installBundle(location, input);
+		bundles.add(bundle);
+		return bundle;
+	}
+
+	@Override
+	public Bundle installBundle(String location) throws BundleException {
+		Bundle bundle = bundleContext.installBundle(location);
+		bundles.add(bundle);
+		return bundle;
+	}
+
+	@Override
+	public Bundle getBundle(long id) {
+		return bundleContext.getBundle(id);
+	}
+
+	@Override
+	public Bundle[] getBundles() {
+		return bundleContext.getBundles();
+	}
+
+	@Override
+	public void addServiceListener(ServiceListener listener, String filter) throws InvalidSyntaxException {
+		bundleContext.addServiceListener(listener, filter);
+		listeners.add(listener);
+	}
+
+	@Override
+	public void addServiceListener(ServiceListener listener) {
+		bundleContext.addServiceListener(listener);
+		listeners.add(listener);
+	}
+
+	@Override
+	public void removeServiceListener(ServiceListener listener) {
+		bundleContext.removeServiceListener(listener);
+		listeners.remove(listener);
+	}
+
+	@Override
+	public void addBundleListener(BundleListener listener) {
+		bundleContext.addBundleListener(listener);
+		listeners.add(listener);
+	}
+
+	@Override
+	public void removeBundleListener(BundleListener listener) {
+		bundleContext.removeBundleListener(listener);
+		listeners.remove(listener);
+	}
+
+	@Override
+	public void addFrameworkListener(FrameworkListener listener) {
+		bundleContext.addFrameworkListener(listener);
+		listeners.add(listener);
+	}
+
+	@Override
+	public void removeFrameworkListener(FrameworkListener listener) {
+		bundleContext.removeFrameworkListener(listener);
+		listeners.remove(listener);
+	}
+
+	@Override
+	public ServiceRegistration<?> registerService(String[] clazzes, Object service, Dictionary<String, ?> properties) {
+		ServiceRegistration<?> reg = bundleContext.registerService(clazzes, service, properties);
+		regs.add(reg);
+		return reg;
+	}
+
+	@Override
+	public ServiceRegistration<?> registerService(String clazz, Object service, Dictionary<String, ?> properties) {
+		ServiceRegistration<?> reg = bundleContext.registerService(clazz, service, properties);
+		regs.add(reg);
+		return reg;
+	}
+
+	@Override
+	public <S> ServiceRegistration<S> registerService(Class<S> clazz, S service, Dictionary<String, ?> properties) {
+		ServiceRegistration<S> reg = bundleContext.registerService(clazz, service, properties);
+		regs.add(reg);
+		return reg;
+	}
+
+	@Override
+	public <S> ServiceRegistration<S> registerService(Class<S> clazz, ServiceFactory<S> factory,
```

Comment:
Not sure why we need to handle services since the framework will already handle cleaning when the bundle in stopped. Or is the point of doing this to deal with cleaning up between test method invocations?

Reply:
but testing is not bundles
all too often we have false positives because the variants of one test leak into other tests in the same class.